### PR TITLE
[SPARK-30220] Enable using Exists/In subqueries outside of the Filter node

### PR DESCRIPTION
**Pull request by tanelk**
### What changes were proposed in this pull request?
Enable using Exists/In subqueries in other nodes besides `Filter`: `Aggregate`, `Project` and `Window`.
This is allready mostly supported, but it was blocked in the `Analyzer`. Only requires a small tweak in the `Optimizer`.

### Why are the changes needed?
One of the last open feature parities between PostgreSQL and Spark: [SPARK-30374](https://issues.apache.org/jira/browse/SPARK-30374)

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing and new UTs

### CODE REVIEW

**Comment by tanelk** [comment]
pinging @viirya , @cloud-fan 
Could you take a look at this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -687,10 +687,10 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
 
       case inSubqueryOrExistsSubquery =>
         plan match {
-          case _: Filter | _: SupportsSubquery | _: Join => // Ok
+          case _: SupportsSubquery | _: Join | _: UnaryNode => // Ok
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:690]: How to make sure that all UnaryNode can support IN/EXISTS subqueries (correlated and uncorrelated)? The test cases only covered Project, Aggregate, and Window. Maybe we could either only support these three nodes for now, or add more tests to cover other unary operators.

Also, would it work with Aggregate having correlated IN/EXISTS subqueries in the GROUP BY clause?

Comment (tanelk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:690]: The Exists/In subqueries in UnaryNodes are handled by inserting an existance join (or joins) bellow the UnaryNode and then the subquery is replaced with an attribute reference. This is a very generic pattern and should not depend on the actual UnaryNode. Yes, Filter is handled separately but that is a performance optimization, not a correctness issue (LeftSemi/LeftAnti are better than existance join).

But if you think that we should stay safe and only allow a subset of UnaryNodes, then I would be totally fine with supporting only the most common nodes - Project, Aggregate, Window and perhaps Sort.

I'll add some extra test cases for aggregate.

Comment (allisonwang-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:690]: I think it's safer to use a whitelist of nodes than to support all UnaryNodes, and it shouldn't be a common use case for EXISTS/IN subquery to be in Generate, Expand, Limit, etc. We can always add more nodes in the future if needed.

**Comment by allisonwang-db** [comment]
cc @cloud-fan

**Comment by tanelk** [comment]
@HyukjinKwon could you take a look at this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 2]
```diff
@@ -782,10 +782,11 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
 
       case inSubqueryOrExistsSubquery =>
         plan match {
-          case _: Filter | _: SupportsSubquery | _: Join => // Ok
+          case _: Filter | _: SupportsSubquery | _: Join |
+            _: Project | _: Aggregate | _: Window => // Ok
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:786]: shall we simply make these plan nodes extend `SupportsSubquery`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala [hunk 3]
```diff
@@ -658,16 +658,6 @@ class AnalysisErrorSuite extends AnalysisTest {
     assertAnalysisError(plan2, "EqualTo does not support ordering on type map" :: Nil)
   }
 
-  test("PredicateSubQuery is used outside of a filter") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:661]: can we still keep this test by putting subquery in sort?

File: sql/core/src/test/resources/sql-tests/inputs/subquery/exists-subquery/exists-outside-filter.sql [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+-- correlated select exist
+-- TC.01.02
+SELECT
+  emp_name,
+  EXISTS (SELECT 1
+          FROM   dept
+          WHERE  emp.dept_id = dept.dept_id)
+FROM   emp;
+
+-- uncorrelated exist in aggregate filter
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/subquery/exists-subquery/exists-outside-filter.sql:55]: do we support subqueries in aggregate function inputs?

Comment (tanelk) [reply sql/core/src/test/resources/sql-tests/inputs/subquery/exists-subquery/exists-outside-filter.sql:55]: Added testcases

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala (hunk 3 above)
Comment (tanelk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:661]: Restrored it

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 2 above)
Comment (tanelk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:786]:
```
/**
 * A trait to represent the commands that support subqueries.
 * This is used to allow such commands in the subquery-related checks.
 */
trait SupportsSubquery extends LogicalPlan
```

Yeah, I'm bit sceptical whit this one. It states, that it works on commands.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**